### DISCUSSION TITLE
Question: How to get a list undefined variables?

### DISCUSSION THREAD

**Discussion by tmtron** [josdejong/mathjs#2865] [Q&A]
e.g. the user may enter a formula like `2 * scale * pi * + offset`
* `pi` is known by MathJs
* but the other 2 variables are unknown  and must be provided by the user when we want to evaluate the expression

How can we easily get a list of all unknown variables?

i.e. we can evaluate the expression with an empty scope and it will throw - but this will only return the first undefined symbol

**Comment by josdejong** [comment]
I think this is a duplicate of #1414. It will take a bit of effort.

**Comment by tmtron** [comment]
@josdejong thanks for the link that makes sense. Maybe the library should contain a convenience function, since this could be useful in many projects.

**Comment by tmtron** [comment]
After some testing, I don't see how #1414 can solve this.

e.g. `math.parse('var*pi');` returns this tree:
```json
{
    "mathjs": "OperatorNode",
    "op": "*",
    "fn": "multiply",
    "args": [
        {
            "mathjs": "SymbolNode",
            "name": "var"
        },
        {
            "mathjs": "SymbolNode",
            "name": "pi"
        }
    ],
}
```

Both nodes `var` and `pi` are `SymbolNodes` with paths `arg[0]`/`arg[1]` and they have the same parent-node (when we traverse the tree).  
So how could we know, that `pi` is a known symbol and `var` is an undefined variable, that the user must provide?

**Comment by josdejong** [comment]
> So how could we know, that `pi` is a known symbol and `var` is an undefined variable, that the user must provide?

You can check whether the symbol is defined or not, for example:

```js
function isDefined(symbol) {
  return math.expression.mathWithTransform[symbol] !== undefined || math.Unit.isValuelessUnit(symbol)
}

isDefined('pi') // true
isDefined('var') // false
isDefined('foo') // false
isDefined('millimeter') // true
```

**Comment by tmtron** [comment]
@josdejong that would also report e.g. `math.AccessorNode` as defined, which is not true

**Reply by josdejong** [reply]
Ah, good point. I've updated [my example](https://github.com/josdejong/mathjs/discussions/2865#discussioncomment-4506013) to use `math.expression.mathWithTransform[symbol]` instead of `math[symbol]`, since the former is the actual object exposed inside the expression parser.
